// @sanity/runtime-cli/dist/commands/blueprints/add.js
import { Args, Command, Flags } from '@oclif/core';
import { blueprintAddCore } from '../../cores/blueprints/index.js';
export default class AddCommand extends Command {
    static description = 'Add a Resource to a Blueprint';
    static examples = [
        '<%= config.bin %> <%= command.id %> function',
        '<%= config.bin %> <%= command.id %> function --name my-function',
        '<%= config.bin %> <%= command.id %> function --name my-function --fn-type document-publish',
        '<%= config.bin %> <%= command.id %> function --name my-function --fn-type document-publish --lang js',
    ];
    static args = {
        type: Args.string({
            description: 'Type of resource to add (e.g. function)',
            options: ['function'],
            required: true,
        }),
    };
    static flags = {
        name: Flags.string({
            description: 'Name of the Resource to add',
            char: 'n',
        }),
        'fn-type': Flags.string({
            description: 'Type of new Function',
            options: ['document-publish' /*, 'document-create', 'document-delete'*/],
            aliases: ['function-type'],
            // default: 'document-publish', // prompting informs user of default
            dependsOn: ['name'],
        }),
        language: Flags.string({
            description: 'Language of the new Function',
            aliases: ['function-language', 'lang'],
            options: ['ts', 'js'],
            default: 'ts',
        }),
        javascript: Flags.boolean({
            description: 'Use JavaScript instead of TypeScript',
            aliases: ['js'],
        }),
    };
    async run() {
        const { args, flags } = await this.parse(AddCommand);
        const { success, error } = await blueprintAddCore({
            bin: this.config.bin,
            log: (msg) => this.log(msg),
            args,
            flags,
        });
        if (!success)
            this.error(error);
    }
}
